File: code/go/ch09/src/luago/state/api_set.go
package state

import . "luago/api"

// [-2, +0, e]
// http://www.lua.org/manual/5.3/manual.html#lua_settable
func (self *luaState) SetTable(idx int) {
	t := self.stack.get(idx)
	v := self.stack.pop()
	k := self.stack.pop()
	self.setTable(t, k, v)
}

// [-1, +0, e]
// http://www.lua.org/manual/5.3/manual.html#lua_setfield
func (self *luaState) SetField(idx int, k string) {
	t := self.stack.get(idx)
	v := self.stack.pop()
	self.setTable(t, k, v)
}

// [-1, +0, e]
// http://www.lua.org/manual/5.3/manual.html#lua_seti
func (self *luaState) SetI(idx int, i int64) {
	t := self.stack.get(idx)
	v := self.stack.pop()
	self.setTable(t, i, v)
}

// [-1, +0, e]
// http://www.lua.org/manual/5.3/manual.html#lua_setglobal
func (self *luaState) SetGlobal(name string) {
	t := self.registry.get(LUA_RIDX_GLOBALS)
	v := self.stack.pop()
	self.setTable(t, name, v)
}

// [-0, +0, e]
// http://www.lua.org/manual/5.3/manual.html#lua_register
func (self *luaState) Register(name string, f GoFunction) {
	self.PushGoFunction(f)
	self.SetGlobal(name)
}

// t[k]=v
func (self *luaState) setTable(t, k, v luaValue) {
	if tbl, ok := t.(*luaTable); ok {
		tbl.put(k, v)
		return
	}

	panic("not a table!")
}

File: code/csharp/ch08/luavm/state/ApiSet.cs
using System;

namespace luavm.state
{
    partial struct LuaState
    {
        public void SetTable(int idx)
        {
            var v = stack.pop();
            var k = stack.pop();
            setTable(idx, new LuaValue(k), new LuaValue(v));
        }

        void setTable(int idx, LuaValue k, LuaValue v)
        {
            var t = new LuaValue(stack.get(idx));
            if (t.isLuaTable())
            {
                var table = t.toLuaTable();
                table.put(k, v);
                stack.set(idx, table);
                return;
            }

            throw new Exception("not a table!");
        }

        public void SetField(int idx, string k)
        {
            var v = stack.pop();
            setTable(idx, new LuaValue(k), new LuaValue(v));
        }

        public void SetI(int idx, long n)
        {
            var v = stack.pop();
            setTable(idx, new LuaValue(n), new LuaValue(v));
        }
    }
}